File: src/main/java/one/pkg/modpublish/api/GithubAPI.java
/*
 * Copyright (C) 2025 404Setup (https://github.com/404Setup)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package one.pkg.modpublish.api;

import com.google.gson.JsonObject;
import com.intellij.openapi.project.Project;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import one.pkg.modpublish.data.internel.ModInfo;
import one.pkg.modpublish.data.internel.PublishData;
import one.pkg.modpublish.data.network.github.GithubData;
import one.pkg.modpublish.data.result.BackResult;
import one.pkg.modpublish.data.result.PublishResult;
import one.pkg.modpublish.data.result.Result;
import one.pkg.modpublish.settings.properties.PID;
import one.pkg.modpublish.util.io.GitInfo;
import one.pkg.modpublish.util.io.JsonParser;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.util.Optional;

public class GithubAPI extends API {
    private static final String RELEASES_URL = "https://api.github.com/repos/{path}/releases";
    private static final String UPLOAD_URL = "https://uploads.github.com/repos/{path}/releases/{id}/assets";
    private static final String REPO_INFO_URL = "https://api.github.com/repos/{path}";
    private static final String BRANCH_COMMIT_URL = "https://api.github.com/repos/{path}/branches/{branch}";

    @NotNull
    public Request.Builder getRequestBuilder(@NotNull String url, @NotNull Project project) {
        return getBaseRequestBuilder().header("Accept", "application/vnd.github+json")
                .header("X-GitHub-Api-Version", "2022-11-28")
                .header("Authorization", "Bearer " + PID.GithubToken.getProtect(project).data())
                .url(url.replace("{path}", PID.GithubRepo.get(project)));
    }

    @Override
    public @NotNull String getID() {
        return "Github";
    }

    @Override
    public void updateABServer() {
    }

    @Override
    public boolean getABServer() {
        return true;
    }

    @Override
    @NotNull
    String createJsonBody(@NotNull PublishData data, @NotNull Project project) {
        String branch = PID.GithubBranch.get(project);
        if (branch.isEmpty()) branch = GitInfo.getBrach(project);
        String targetCommitish = getTargetCommitish(branch, project);

        GithubData.GithubDataBuilder githubData = GithubData.builder()
                .tagName(data.versionNumber().startsWith("v") ?
                        data.versionNumber() :
                        "v" + data.versionNumber())
                .targetCommitish(targetCommitish)
                .name(data.versionName())
                .body(data.changelog())
                .releaseChannel(data.releaseChannel())
                .draft(false)
                .generateReleaseNotes(true)
                .makeLatest(true);

        return githubData.build().toJson();
    }

    @Override
    public @NotNull PublishResult createVersion(@NotNull PublishData data, @NotNull Project project) {
        try {
            String tagName = data.versionNumber().startsWith("v") ?
                    data.versionNumber() :
                    "v" + data.versionNumber();

            Optional<JsonObject> existingRelease = checkExistingRelease(tagName, project);

            JsonObject releaseResponse;
            if (existingRelease.isPresent()) {
                releaseResponse = existingRelease.get();
            } else {
                Result releaseResult = createRelease(data, project);
                if (releaseResult instanceof PublishResult pr) {
                    return pr;
                }
                BackResult br = (BackResult) releaseResult;
                releaseResponse = JsonParser.fromJson(br.asString()).getAsJsonObject();
            }

            String uploadUrl = releaseResponse.get("upload_url").getAsString();
            uploadUrl = uploadUrl.split("\\{")[0];

            PublishResult result = PublishResult.empty();
            for (File file : data.files()) {
                result = uploadAsset(file, project, uploadUrl);
                if (!result.isSuccess()) break;
            }

            return result;

        } catch (Exception e) {
            return PublishResult.create(this, "Failed to create GitHub release: " + e.getMessage());
        }
    }

    @NotNull
    private Result createRelease(@NotNull PublishData data, @NotNull Project project) {
        try {
            Request request = getJsonRequest(getRequestBuilder(RELEASES_URL, project))
                    .post(RequestBody.create(createJsonBody(data, project), MediaType.get("application/json")))
                    .build();

            try (Response response = NetworkUtil.client.newCall(request).execute()) {
                Optional<String> status = getStatus(response);
                if (status.isPresent()) return PublishResult.create(this, "Failed to create release: " + status.get());
                return BackResult.result(response.body().string());
            }
        } catch (IOException e) {
            return PublishResult.create(this, "Network error: " + e.getMessage());
        }
    }

    @NotNull
    private PublishResult uploadAsset(@NotNull File file, @NotNull Project project, @NotNull String uploadUrl) {
        try {
            String fileName = file.getName();
            String assetUrl = uploadUrl + "?name=" + fileName;

            RequestBody fileBody = RequestBody.create(file, MediaType.get("application/java-archive"));

            Request request = getRequestBuilder(assetUrl, project)
                    .header("Content-Type", "application/java-archive")
                    .post(fileBody)
                    .build();

            try (Response response = NetworkUtil.client.newCall(request).execute()) {
                Optional<String> status = getStatus(response);
                /*JsonObject assetResponse = JsonParser.getJsonObject(response.body().byteStream());
                String downloadUrl = assetResponse.get("browser_download_url").getAsString();*/
                return status.map(s -> PublishResult.create(this, "Failed to upload asset: " + s)).orElseGet(PublishResult::empty);
            }
        } catch (IOException e) {
            return PublishResult.create(this, "Failed to upload asset: " + e.getMessage());
        }
    }

    @NotNull
    private Optional<JsonObject> checkExistingRelease(@NotNull String tagName, @NotNull Project project) {
        try {
            String url = RELEASES_URL + "/tags/" + tagName;
            Request request = getRequestBuilder(url, project)
                    .get()
                    .build();

            try (Response response = NetworkUtil.client.newCall(request).execute()) {
                if (response.isSuccessful()) {
                    String responseBody = response.body().string();
                    JsonObject releaseInfo = JsonParser.fromJson(responseBody);
                    return Optional.of(releaseInfo);
                } else if (response.code() == 404) {
                    return Optional.empty();
                } else {
                    return Optional.empty();
                }
            }
        } catch (IOException e) {
            return Optional.empty();
        }
    }

    @NotNull
    private String getTargetCommitish(@NotNull String branch, @NotNull Project project) {
        try {
            if (branch.trim().isEmpty()) {
                return getDefaultBranch(project);
            } else {
                return getLatestCommitHash(branch, project);
            }
        } catch (Exception e) {
            return !branch.trim().isEmpty() ? branch : "main";
        }
    }

    @NotNull
    private String getDefaultBranch(@NotNull Project project) throws IOException {
        Request request = getRequestBuilder(REPO_INFO_URL, project)
                .get()
                .build();

        try (Response response = NetworkUtil.client.newCall(request).execute()) {
            if (response.isSuccessful()) {
                JsonObject repoInfo = JsonParser.fromJson(response.body().string());
                return repoInfo.get("default_branch").getAsString();
            }
        }
        return "main";
    }

    @NotNull
    private String getLatestCommitHash(@NotNull String branch, @NotNull Project project) throws IOException {
        String url = BRANCH_COMMIT_URL.replace("{branch}", branch);
        Request request = getRequestBuilder(url, project)
                .get()
                .build();

        try (Response response = NetworkUtil.client.newCall(request).execute()) {
            if (response.isSuccessful()) {
                JsonObject branchInfo = JsonParser.fromJson(response.body().string());
                JsonObject commit = branchInfo.getAsJsonObject("commit");
                return commit.get("sha").getAsString();
            }
        }
        return branch;
    }


    @Override
    public @NotNull ModInfo getModInfo(@NotNull String modid, @NotNull Project project) {
        return ModInfo.empty();
    }
}

File: src/main/kotlin/one/pkg/modpublish/api/GithubAPI.kt
/*
 * Copyright (C) 2025 - 2026 404Setup (https://github.com/404Setup)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package one.pkg.modpublish.api

import com.google.gson.JsonObject
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import one.pkg.modpublish.api.NetworkUtil.client
import one.pkg.modpublish.data.internal.ModInfo
import one.pkg.modpublish.data.internal.PublishData
import one.pkg.modpublish.data.network.github.GithubData
import one.pkg.modpublish.data.result.BackResult
import one.pkg.modpublish.data.result.PublishResult
import one.pkg.modpublish.data.result.Result
import one.pkg.modpublish.settings.properties.PID
import one.pkg.modpublish.util.io.GitInfo.getBrach
import one.pkg.modpublish.util.io.JsonParser.fromJson
import one.pkg.modpublish.util.io.JsonParser.toJson
import java.io.File
import java.io.IOException

class GithubAPI : GitPlatformAPI() {
    override val id: String = "Github"

    override suspend fun createVersion(data: PublishData, project: Project): PublishResult {
        return runCatching {
            val tagName = if (data.versionNumber.startsWith("v")) data.versionNumber else "v${data.versionNumber}"
            val existingRelease = checkExistingRelease(tagName, project)?.asJsonObject ?: run {
                val result = createRelease(data, project)
                if (result is PublishResult) return result
                (result as BackResult).asString().fromJson().asJsonObject
            }

            val uploadUrl =
                existingRelease.get("upload_url")?.asString?.split("{")?.first() ?: return PublishResult.create(
                    this,
                    "Failed to get upload URL"
                )
            data.files.fold(PublishResult.EMPTY) { acc, file ->
                if (!acc.isSuccess) acc else uploadAsset(file, project, uploadUrl)
            }
        }.getOrElse { PublishResult.create(this, "Failed to create GitHub release: ${it.message}") }
    }

    private fun HttpRequestBuilder.github(project: Project) {
        header("Accept", "application/vnd.github+json")
        header("X-GitHub-Api-Version", "2022-11-28")
        header("Authorization", "Bearer ${PID.GithubToken.getProtect(project).data}")
    }

    private suspend fun createRelease(data: PublishData, project: Project): Result = runCatching {
        val repo = PID.GithubRepo.get(project)
        val resp = client.post(RELEASES_URL.replace("{path}", repo)) {
            github(project)
            contentType(ContentType.Application.Json)
            //json()
            setBody(createJsonBodyAsync(data, project))
        }

        resp.statusString()?.let { return PublishResult.create(this, "Failed to create release: $it") }
        BackResult.result(resp.bodyAsText())
    }.getOrElse { PublishResult.create(this, "Network error: ${it.message}") }

    private suspend fun uploadAsset(file: File, project: Project, uploadUrl: String): PublishResult = runCatching {
        val assetUrl = "$uploadUrl?name=${file.name}"
        val resp = client.post(assetUrl) {
            github(project)
            header("Content-Type", "application/java-archive")
            setBody(file.readBytes())
        }

        resp.statusString()?.let { return PublishResult.create(this, "Failed to upload asset: $it") }
        PublishResult.EMPTY
    }.getOrElse { PublishResult.create(this, "Failed to upload asset: ${it.message}") }

    private suspend fun checkExistingRelease(tagName: String, project: Project): JsonObject? = runCatching {
        val url = "$RELEASES_URL/tags/$tagName".replace("{path}", PID.GithubRepo.get(project))
        val resp = client.get(url) { github(project) }
        if (resp.status.isSuccess()) resp.bodyAsText().fromJson() else null
    }.getOrNull()

    @Throws(IOException::class)
    override suspend fun getDefaultBranch(project: Project): String {
        val repo = PID.GithubRepo.get(project)
        val resp = client.get(REPO_INFO_URL.replace("{path}", repo)) { github(project) }
        if (resp.status.isSuccess()) return resp.bodyAsText().fromJson().get("default_branch").asString
        return "main"
    }

    @Throws(IOException::class)
    override suspend fun getLatestCommitHash(branch: String, project: Project): String {
        val repo = PID.GithubRepo.get(project)
        val url = BRANCH_COMMIT_URL.replace("{path}", repo).replace("{branch}", branch)
        val resp = client.get(url) { github(project) }
        if (resp.status.isSuccess()) {
            val commit = resp.bodyAsText().fromJson().getAsJsonObject("commit")
            return commit.get("sha").asString
        }
        return branch
    }

    override suspend fun getModInfo(modid: String, project: Project): ModInfo = ModInfo.empty()

    override fun createJsonBody(data: PublishData, project: Project): String {
        return ""
    }

    private suspend fun createJsonBodyAsync(data: PublishData, project: Project): String {
        val branch = PID.GithubBranch.get(project).ifEmpty { project.getBrach() }
        val targetCommitish = getTargetCommitish(branch, project)

        return GithubData().apply {
            this.tagName = if (data.versionNumber.startsWith("v")) data.versionNumber else "v${data.versionNumber}"
            this.targetCommitish = targetCommitish
            this.name = data.versionName
            this.body = data.changelog
            this.releaseChannel(data.releaseChannel)
            this.draft = false
            this.generateReleaseNotes = true
            this.makeLatest(true)
        }.toJson().apply {
            LOG.info("now run $id publish: $this")
        }
    }

    companion object {
        private const val RELEASES_URL = "https://api.github.com/repos/{path}/releases"
        private const val REPO_INFO_URL = "https://api.github.com/repos/{path}"
        private const val BRANCH_COMMIT_URL = "https://api.github.com/repos/{path}/branches/{branch}"
        private val LOG = Logger.getInstance(GithubAPI::class.java)
    }
}